rcl/test/rcl/test_node.cpp

```diff
@@ -351,3 +380,145 @@ TEST_F(CLASSNAME(TestNodeFixture, RMW_IMPLEMENTATION), test_rcl_node_life_cycle)
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAME, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node name with /, which is valid in a topic, but not a node name.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, "my/node_42", namespace_, &default_options);
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAME, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node name with {}, which is valid in a topic, but not a node name.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, "my_{node}_42", namespace_, &default_options);
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAME, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+}
+
+/* Tests the node namespace restrictions enforcement.
+ */
+TEST_F(CLASSNAME(TestNodeFixture, RMW_IMPLEMENTATION), test_rcl_node_namespace_restrictions) {
+  stop_memory_checking();
+  rcl_ret_t ret;
+
+  // Initialize rcl with rcl_init().
+  ret = rcl_init(0, nullptr, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret);
+  auto rcl_shutdown_exit = make_scope_exit([]() {
+    rcl_ret_t ret = rcl_shutdown();
+    ASSERT_EQ(RCL_RET_OK, ret);
+  });
+
+  const char * name = "node";
+  rcl_node_options_t default_options = rcl_node_get_default_options();
+
+  // First do a normal node namespace.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, name, "/ns", &default_options);
+    ASSERT_EQ(RCL_RET_OK, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node namespace which is an empty string, which is also valid.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, name, "", &default_options);
+    ASSERT_EQ(RCL_RET_OK, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node namespace which is just a forward slash, which is invalid.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, name, "/", &default_options);
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAMESPACE, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node namespaces with invalid characters.
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, name, "/ns/{name}", &default_options);
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAMESPACE, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+  {
+    rcl_node_t node = rcl_get_zero_initialized_node();
+    ret = rcl_node_init(&node, name, "/~/", &default_options);
+    ASSERT_EQ(RCL_RET_NODE_INVALID_NAMESPACE, ret);
+    rcl_ret_t ret = rcl_node_fini(&node);
+    EXPECT_EQ(RCL_RET_OK, ret);
+  }
+
+  // Node namespace with a trailing / which is not allowed.
```

Comment:
I'd rather leave this kind of flexibility to the top level API's (like rclcpp and rclpy). Each corner case makes it harder to describe and harder to test, so I'd like to keep `rcl` simple by avoiding them. That's just my personal preference though. I have no problem with allowing this exception to the rule here or in the higher level API's.